### ISSUE TITLE
Deprecate zlib support

### ISSUE THREAD

**Issue by paulidale** [openssl/openssl#14506]
From: #14499

Deprecate everything to do with zlib support.

**Comment by richsalz** [comment]
If the OTC approves, I volunteer to take this on.

**Comment by dcooper16** [comment]
Would *everything* include use of zlib in TLS certificate compression, RFC 8879 (https://github.com/openssl/openssl/pull/9155)?

**Comment by t8m** [comment]
Not necessarily - however the TLS certificate compression support is not going to be included in 3.0 anyway. If something zlib related is deprecated and would be needed for the RFC 8879 support when it is merged, it can be easily undeprecated.

**Comment by levitte** [comment]
The zlib support is quite messy anyway.  I'd be happy to see it replaced with something better.

**Comment by paulidale** [comment]
More specifically:

- deprecate the functions in `<openssl/comp.h>`, OpenSSL is not a compression library and shouldn't be providing this API.
- deprecate the public SSL functions dealing with compression, it shouldn't be used.
- remove the `-z` option from the `enc` command, it compresses the cipher text which isn't ideal.

**Comment by richsalz** [comment]
This is still waiting for the OTC to officially decide, right?

**Comment by mattcaswell** [comment]
OTC are ok with this happening if there is a PR.

**Comment by tmshort** [comment]
Well... that changes certificate compression, then, doesn't it?

**Comment by richsalz** [comment]
> Well... that changes certificate compression, then, doesn't it?

the deprecation will have to be done carefully.  i was thinking of another config option that would set/clear `OPENSSL_NO_RFC8879` that would interact with the `OPENSSL_NO_COMPRESSION` flag.

**Comment by tmshort** [comment]
Personally, I'd prefer to get #9155 merged before this happens...

**Comment by t8m** [comment]
Apparently this will be harder than expected because the COMP api is too much intertwined in libssl and proper deprecation would mean changing apps and in libssl moving everything COMP related to the tls_depr sources. So probably not a 3.0 material.

**Comment by t8m** [comment]
The milestone was supposed to be removed during the Mar 23rd OTC meeting.

**Comment by richsalz** [comment]
I'm going to try and handle this in smaller PR's. First, remove SSL/TLS compression (done in #15621). Then perhaps `BIO_f_zlib()`

**Comment by tmshort** [comment]
@richsalz anything you do to remove BIO compression will impact certificate compression.
Removing from TLS/SSL should not impact cert compression

**Comment by richsalz** [comment]
Yes, but that API is a piece of gunk.  I think cert compression should have a callback that gives the whole cert, and gets back the compressed form.  Perhaps some registration with ID's also.

**Comment by slontis** [comment]
> Yes, but that API is a piece of gunk. I think cert compression should have a callback that gives the whole cert, and gets back the compressed form. Perhaps some registration with ID's also.

This is sounding like a post 3.0 item, since this is not sounding like just doing a simple deprecation

**Comment by t8m** [comment]
I think at most the SSL_COMP could be deprecated in 3.0. Deprecating the BIO_f_zlib() would be too big change to make it into 3.0.

**Comment by richsalz** [comment]
Certificate compression isn't in 3.0, so the fact that the PR is currently written to use `BIO_f_zlib` should not matter much.

The PR is there, #15625, y'all can decide if it's too big for 3.0  (It *is* just anot compress-related deprecation)

**Comment by t8m** [comment]
Closing in favor of #28348
